lib/core/src/Cardano/Pool/Ranking.hs

```diff
@@ -0,0 +1,198 @@
+-- +------------+-------------+------------------------------------------------+
+module Cardano.Pool.Ranking
+    (
+      -- * Formulas
+      desirability
+
+    , saturatedPoolRewards
+    , saturatedPoolSize
+
+      -- * Types
+    , EpochConstants (..)
+    , Pool (..)
+    , Lovelace (..)
+    , Ratio
+    , unsafeToRatio
+    , getRatio
+    , unsafeMkRelativeStake
+    , NonNegative (..)
+    , Positive (..)
+    , unsafeToPositive
+    , unsafeToNonNegative
+    )
+    where
+
+import Prelude
+
+import Data.Word
+    ( Word64 )
+import GHC.Generics
+    ( Generic )
+
+--------------------------------------------------------------------------------
+-- Formulas from spec
+--------------------------------------------------------------------------------
+
+-- | Non-myopic pool desirability according to section 5.6.1.
+--
+-- Is /not/ affected by oversaturation nor pool stake in general.
+desirability
+    :: EpochConstants
+    -> Pool
+    -> Double
+desirability constants pool
+    | f_saturated <= c = 0
+    | otherwise    = (f_saturated - c) * (1 - m)
+  where
+    f_saturated = saturatedPoolRewards constants pool
+    m = getRatio $ margin pool
+    c = fromIntegral $ getLovelace $ cost pool
+
+-- | Total rewards for a pool if it were saturated.
+--
+-- When a0 = 0 this reduces to just p*R*z0 (tested
+-- by @prop_saturatedPoolRewardsReduces@)
+saturatedPoolRewards :: EpochConstants -> Pool -> Double
+saturatedPoolRewards constants pool =
+    let
+        a0 = getNonNegative $ leaderStakeInfluence constants
+        z0 = getRatio $ saturatedPoolSize constants
+        s = getRatio $ leaderStake pool
+        _R = fromIntegral $ getLovelace $ totalRewards constants
+        p = getNonNegative $ recentAvgPerformance pool
+        -- ^ technically \hat{p} in the spec
+    in
+        (p * _R) / (1 + a0)
+        * (z0 + ((min s z0) * a0))
+
+-- | Determines z0, i.e 1 / k
+saturatedPoolSize :: EpochConstants -> Ratio
+saturatedPoolSize constants =
+    Ratio $ 1 / fromIntegral (getPositive $ desiredNumberOfPools constants)
+
+--------------------------------------------------------------------------------
+-- Types
+--------------------------------------------------------------------------------
+
+data EpochConstants = EpochConstants
+    { leaderStakeInfluence :: NonNegative Double
+      -- ^ a_0
+    , desiredNumberOfPools :: Positive Int
+      -- ^ k
+    , totalRewards :: Lovelace
+      -- ^ Total rewards in an epoch. "R" in the spec.
+    } deriving (Show, Eq, Generic)
+
+data Pool = Pool
+    { leaderStake :: Ratio
+      -- ^ s
+    , cost :: Lovelace
+      -- ^ c
+    , margin :: Ratio
+      -- ^ m
+    , recentAvgPerformance :: NonNegative Double
+      -- ^ \hat{p}, an already averaged performance-value.
+      --
+      -- Should mostly be in the range [0, 1], but lucky pools may exceed 1.
+    } deriving (Show, Eq, Generic)
+
+newtype Lovelace = Lovelace { getLovelace :: Word64 }
```

Comment:
Minor note: I find "Natural" conceptually better than `Word64` (in this context at least, but generally speaking) to describe unbounded positive integers. `Word64` is something we would find at boundaries of the system like databases or types in network serializations. Yet, conceptually, the _right_ type to describe positive integers is `Natural`.